app/src/main/java/ru/karasevm/privatednstoggle/service/ShortcutService.kt

```diff
@@ -0,0 +1,73 @@
+package ru.karasevm.privatednstoggle.service
+
+import android.app.Service
+import android.content.Intent
+import android.os.IBinder
+import android.widget.Toast
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import ru.karasevm.privatednstoggle.PrivateDNSApp
+import ru.karasevm.privatednstoggle.data.DnsServerRepository
+import ru.karasevm.privatednstoggle.util.PreferenceHelper
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils.DNS_MODE_AUTO
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils.DNS_MODE_OFF
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils.DNS_MODE_PRIVATE
+import ru.karasevm.privatednstoggle.util.PrivateDNSUtils.checkForPermission
+
+class ShortcutService : Service() {
+
+    private val repository: DnsServerRepository by lazy { (application as PrivateDNSApp).repository }
+    private val job = SupervisorJob()
+    private val scope = CoroutineScope(Dispatchers.IO + job)
+
+    companion object {
+        private const val ACTION_SWITCH_MODE = "privatednstoggle://switch_mode"
+    }
+
+    private fun setDnsModeAndShowToast(dnsMode: String) {
+        if(dnsMode == DNS_MODE_PRIVATE) {
+            val provider = PrivateDNSUtils.getPrivateProvider(contentResolver)
+            if (provider == null) {
+                PrivateDNSUtils.setPrivateProvider(contentResolver, "dns.google")
+            }
+        }
+        PrivateDNSUtils.setPrivateMode(contentResolver, dnsMode)
+        val text = when (dnsMode) {
```

Comment:
Please don't hardcode strings. We have `set_to_off_toast`, `set_to_auto_toast`, and `set_to_provider_toast` already.